File: tictactoe.py
import random
import sys

class Tictactoe:
    """
    Game class of TicTacToe
    """
    def __init__(self) -> None:
        self.board = [' '] * 9
        # A list of 9 strings, one for each cell, 
        # will contain ' ' or 'X' or 'O'
        self.played = set()    # A set to keep track of the played cells 
        #`win_sets` contains a list of sets in which,
        # all the 3 values within each set must be a subset of 
        # the indices occupied by `who`. 
        # The way this code is laid out ensures that X and O will
        # be able to use this same function.
        self.win_sets = [
            {0,1,2}, # row configuration subsets to win
            {3,4,5},
            {6,7,8},
            {0,3,6}, # cols
            {1,4,7},
            {2,5,8},
            {0,4,8}, # diagonals
            {2,4,6}
        ]
        self.corner_cells = [0,2,6,8]

        self.gui = Screen()
        self.gui.displayWelcomeScreen()
        # Deciding first move 
        if(bool(random.randint(0,1))): # Chooses a random number between 1 and 0 (Heads or Tails)
            self.computer_first_turn = True
            self.gui.displayMessage("Heads. I go first!")
            self.computerNextMove()
        else:
            self.computer_first_turn = False
            self.gui.displayMessage("Tails. you go first!")
        # Show the board for the first time
        self.gui.displayBoard(self.board)    
    
        
    def playerNextMove(self) -> None:
        """ Method that prompts the player for a valid cell number, 
            and prints the info and the updated board;
            error checks that the input is a valid cell number 
        """
        while True: # Keep trying until user enters an integer type
            try:
                raw = self.gui.getPlayerInput()
                cellnum = int(raw)
            except ValueError:  
                self.gui.displayInputError(raw, 'type')
            else:
                if(cellnum < 0 or cellnum > 8):
                    self.gui.displayInputError(cellnum, 'range')

                elif(cellnum in self.played):
                    self.gui.displayInputError(cellnum, 'occupied')

                else:
                    self.gui.displayMessage(f"You chose cell {cellnum}")
                    break
        
        self.played.add(cellnum)
        self.board[cellnum] = 'X'
        self.gui.displayBoard(self.board)


    def hasWon(self, who: str) -> bool:
        """ returns True if who (being passed 'X' or 'O') has won, False otherwise """
        win_indices = [i for i, c in enumerate(self.board) if c == who] 

        if any(oneset.issubset(set(win_indices)) for oneset in self.win_sets):
            return True
        else:
            return False
    

    def empty_cells(self) -> list:
        # Find playable positions based on current unoccupied cells:
        return [i for i, cell in enumerate(self.board) if cell == " "]
    

    def minimax(self, position : int, maximizingPlayer : bool) -> tuple[int, int]:
        """
        This is a minimax algorithm implementation
        that recursively calls itself.
        The computer is the maximizing player, and the algorithm 
        assumes the player to play the best they can

        The computer is the maximizing player, returns the best position to play
        therefore, O is the maximizing player
        
        Args: position (int): The playable position to evaluate a score out of
            maximizingPlayer (bool): Set to false if the computer tries to predict a player move

        Returns:
            max/min_score (int): The highest score witnessed out of this move and all its descendant moves
            best_pos (int) : The best position
        """
        # First, check if last recursive call ended in a win/tie
        if self.hasWon('O'): 
            return 1*len(self.empty_cells()) + 1, position
        elif self.hasWon('X'):
            return -(1*len(self.empty_cells()) + 1), position
        elif len(self.empty_cells()) == 0:
            return 0, position
                
        if maximizingPlayer:
            max_score = -float('inf')
            best_pos = None
            for cell in self.empty_cells():
                self.board[cell] = 'O' # Test a move
                score, _ = self.minimax(cell, False)
                self.board[cell] = ' ' 
                if(score > max_score):
                    max_score = score
                    best_pos = cell # equate to current cell being eval'd if it's the best cell
            return max_score, best_pos
        else:
            min_score = float('inf')
            for cell in self.empty_cells():
                self.board[cell] = 'X' # Fake a player move
                score, _ = self.minimax(cell, True)
                self.board[cell] = ' '
                if(score < min_score):
                    min_score = score # cell corresponding to min_score if therefore what a 'perfect player' would play
                    best_pos = cell
            return min_score, best_pos # technically 'worst pos', but we're in player's POV rn
      

    def computerNextMove(self) -> None:
        """ Computer randomly chooses a valid cell, 
            and prints the info and the updated board 
        """
        cell : int
        temp_cell : int
        temp_score : int
        best_cell = 0
        best_score = -10000
        
        if(len(self.empty_cells()) == 9): # This function was called when the board was empty
            # This means that it's the computer's first move!
            best_cell = random.choice(self.corner_cells) 
        else:
            for cell in self.empty_cells():
                temp_score, temp_cell = self.minimax(cell, True)
                if(temp_score > best_score):
                    best_cell = temp_cell
        
        self.played.add(best_cell) # Adds this chosen cell into the set of occupied cells
        self.board[best_cell] = 'O'
        self.gui.displayBoard(self.board)
    
    def terminate(self, who: str) -> bool:
        """ returns True if who (being passed 'X' or 'O') has won or if it's a draw, False otherwise;
            it also prints the final messages:
                    "You won! Thanks for playing." or 
                    "You lost! Thanks for playing." or 
                    "A draw! Thanks for playing."  
        """
        if(self.hasWon(who) and who == 'X' ):
            self.gui.displayMessage("You won! Thanks for playing.")
            return True
        elif(self.hasWon(who) and who == 'O' ):
            self.gui.displayMessage("You lost! Thanks for playing.")
            return True
        elif(len(self.played) == 9):
            self.gui.displayMessage("A draw! Thanks for playing.")
            return True
        else:
            return False
    

class Screen():
    """
    Fake GUI
    Some screen formatting to make the game look a lot neater.
    ANSI codes are used to set cursor position and to clear the screen or row.
    """
    def __init__(self) -> None:
        self.clearScreen()
        # self.displayWelcomeScreen()
        self.info_message_row = 22
        self.state_messsage_row = 24
        self.board_start_row = 14
        self.player_input_row = 20


    def displayWelcomeScreen(self) -> None:
        self.clearScreen()
        """ prints the banner messages 
            and prints the intial board on the screen
        """
        print("Welcome to Tic-Tac-Toe!\n")
        print("You play X and the computer plays O.\n")
        print("\033[0;96mI've become sentient...\nI, as the computer, will now play strategically!\033[0m\n")
        print("Let's determine who goes first! I call heads.\n")
        input("Press [Enter] to flip a coin ...")


    def displayMessage(self, s: str) -> None:
        self.eraseRow(self.info_message_row) 
        self.setCursor(self.info_message_row,0)
        print(s)


    def displayGameStatus(self, s: str) -> None:
        self.eraseRow(self.state_messsage_row) 
        self.setCursor(self.state_messsage_row,0)
        print(s)


    def displayBoard(self, board: list) -> None:
        self.setCursor(self.board_start_row,0)
        board_print = (f'   {board[0]} | {board[1]} | {board[2]}    0 | 1 | 2\n'
                        f'   --+---+--    --+---+--\n'
                        f'   {board[3]} | {board[4]} | {board[5]}    3 | 4 | 5\n'
                        f'   --+---+--    --+---+--\n'
                        f'   {board[6]} | {board[7]} | {board[8]}    6 | 7 | 8'   
        )
        print(board_print)


    def clearScreen(self) -> None:
        print("\033[2J")
        print("\033[H")
    

    def setCursor(self, row = 0, col = 0) -> None:
        print(f"\033[{row};{col}H")
    

    def eraseRow(self, row = 0) -> None:
        print(f"\033[{row};0H")
        print(f"\033[2K")


    def getPlayerInput(self):
        self.eraseRow(self.player_input_row) # Refreshes the player input line so the last input clears
        self.setCursor(self.player_input_row,0)
        raw = input("> Next move for X (state a valid cell num): ")
        return raw


    def displayInputError(self, input, mode: str) -> None:

        if mode == 'range':
            self.setCursor(self.info_message_row,0)
            print(f"\033[91m '{input}' is occupied! Try another valid cell number\033[0m")

        elif mode == 'type':
            self.setCursor(self.info_message_row,0) 
            print(f"\033[91m '{input}' is not an Integer! Please try again...\033[0m")

        elif mode == 'occupied':
            self.setCursor(self.info_message_row,0) 
            print(f"\033[91m '{input}' is occupied! Try another valid cell number\033[0m")



if __name__ == "__main__":
    
    # Use as is. 
    game = Tictactoe()
    while True:
        while True:
            game.playerNextMove()            # X starts first
            if(game.terminate('X')): break   # if X won or a draw, print message and terminate
            game.computerNextMove()          # computer plays O
            if(game.terminate('O')): break   # if O won or a draw, print message and terminate
        
        game.gui.displayGameStatus("Press [1] to continue game. Press anything else to end.")
        option = input("> ")
        if option == '1':
            game = Tictactoe()
            continue
        else:
            break

    sys.exit()

File: tictactoe.cpp
#include <iostream>
#include <vector>
#include <array>
#include <set>
#include <string>
#include <random>
#include <limits>
#include <algorithm>


class Screen 
{
public:
    Screen() {
        clearScreen();
        info_message_row = 22;
        state_message_row = 24;
        board_start_row   = 14;
        player_input_row  = 20;
    }

    void displayWelcomeScreen() {
        clearScreen();
        std::cout << "Welcome to Tic-Tac-Toe!\n\n";
        std::cout << "You play X and the computer plays O.\n\n";
        std::cout << "\033[0;96mI've become sentient...\nI, as the computer, will now play strategically!\033[0m\n\n";
        std::cout << "Let's determine who goes first! I call heads.\n\n";
        std::cout << "Press [Enter] to flip a coin ...";
        std::string option;
        std::getline(std::cin, option);
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }

    void displayMessage(const std::string &s) {
        eraseRow(info_message_row);
        setCursor(info_message_row, 0);
        std::cout << s << "\n";
    }

    void displayGameStatus(const std::string &s) {
        eraseRow(state_message_row);
        setCursor(state_message_row, 0);
        std::cout << s << "\n";
    }

    void displayBoard(const std::vector<char> &board) {
        setCursor(board_start_row, 0);
        std::cout << "   " << board[0] << " | " << board[1] << " | " << board[2] << "    0 | 1 | 2\n"
                  << "   --+---+--    --+---+--\n"
                  << "   " << board[3] << " | " << board[4] << " | " << board[5] << "    3 | 4 | 5\n"
                  << "   --+---+--    --+---+--\n"
                  << "   " << board[6] << " | " << board[7] << " | " << board[8] << "    6 | 7 | 8\n";
    }

    std::string getPlayerInput() {
        eraseRow(player_input_row);
        setCursor(player_input_row, 0);
        std::cout << "> Next move for X (state a valid cell num): ";
        std::cout.flush();

        std::string raw;
        std::getline(std::cin, raw);
        return raw;
    }

    void displayInputError(const std::string &input, const std::string &mode) {
        setCursor(info_message_row,0);
        if (mode == "range") {
            std::cout << "\033[91m '" << input << "' is not in range 0-8! Please try again...\033[0m\n";
        } else if (mode == "type") {
            std::cout << "\033[91m '" << input << "' is not an Integer! Please try again...\033[0m\n";
        } else if (mode == "occupied") {
            std::cout << "\033[91m '" << input << "' is occupied! Try another valid cell number\033[0m\n";
        }
    }

    void clearScreen() {
        std::cout << "\033[2J" << "\033[H";
    }

    void setCursor(int row = 0, int col = 0) {
        std::cout << "\033[" << row << ";" << col << "H";
    }

    void eraseRow(int row = 0) {
        std::cout << "\033[" << row << ";0H";
        std::cout << "\033[2K";
    }

private:
    int info_message_row;
    int state_message_row;
    int board_start_row;
    int player_input_row;
};


class TicTactoe 
{
public:
    TicTactoe() { // init

        board = std::vector<char>(9, ' ');

        win_sets = {
            std::array<int,3>{0,1,2}, // rows
            std::array<int,3>{3,4,5},
            std::array<int,3>{6,7,8},
            std::array<int,3>{0,3,6}, // cols
            std::array<int,3>{1,4,7},
            std::array<int,3>{2,5,8},
            std::array<int,3>{0,4,8}, // diagonals
            std::array<int,3>{2,4,6}
        };
        num_win_sets = win_sets.size();

        corner_cells = {0,2,6,8};

        bool computer_goes_first = randomBool();

        if (computer_goes_first) {
            computer_first_turn = true;
            gui.displayMessage("Heads. I go first!");
            computerNextMove();
        } else {
            computer_first_turn = false;
            gui.displayMessage("Tails. you go first!");
        }

        gui.displayBoard(board);
    }

    void playerNextMove() {
        while (true) {
            std::string raw = gui.getPlayerInput();
            int cellnum;
            try {
                cellnum = std::stoi(raw);
            } catch (...) {
                gui.displayInputError(raw, "type");
                continue;
            }

            if (cellnum < 0 || cellnum > 8) {
                gui.displayInputError(std::to_string(cellnum), "range");
            } 
            else if (played.find(cellnum) != played.end()) {
                gui.displayInputError(std::to_string(cellnum), "occupied");
            }
            else {
                gui.displayMessage("You chose cell " + std::to_string(cellnum));
                played.insert(cellnum);
                board[cellnum] = 'X';
                gui.displayBoard(board);
                break;
            }
        }
    }
    
    bool hasWon(char who) {
        // scan through the board and get current indices
        std::vector<int> current_indices;
        for (int i = 0; i < 9; i++) {
            if (board[i] == who) { 
                current_indices.push_back(i); // push_back = append
            }
        }

        // see if any set in win_sets is a subset of current_indices
        // convert current_indices to a set, and check for subsets
        std::set<int> current_set(current_indices.begin(), current_indices.end());

        for (auto &wset : win_sets) {
            std::set<int> check_set(wset.begin(), wset.end());
            bool is_subset = true; // "nested" flag
            // this can be improved to quit as soon as a winning set is found
            for (int set : check_set) {
                if (current_set.find(set) == current_set.end()) {
                    is_subset = false;
                    break;
                }
            }
            if (is_subset) return true;
        }
        return false;
    }

    /*
    Returns vector of current empty cells
    */
    std::vector<int> empty_cells() {
      
        std::vector<int> empties;
        for (int i = 0; i < 9; i++) {
            if (board[i] == ' ') {
                empties.push_back(i);
            }
        }
        return empties;
    }

    std::pair<int, int> minimax(int position, bool maximizingPlayer) {

        // check if last recursive call ended in a win/tie
        int highest_score;
        // conditions to break out of current recursive call
        if (hasWon('O')) {
            highest_score = 1* (int)empty_cells().size() + 1;
            return std::make_pair(highest_score, position);
        } 
        else if(hasWon('X')) {
            highest_score = -1 * ((int)empty_cells().size() + 1);
            return std::make_pair(highest_score, position);
        }
        else if(empty_cells().empty()) {
            return std::make_pair(0, position);
        }

        if (maximizingPlayer) {
            int max_score = -9999999;
            int best_pos = -1;
            for (int cell : empty_cells()) {
                board[cell] = 'O';
                auto result = minimax(cell, false); // make recursive call
                board[cell] = ' ';
                int score = result.first;
                if (score > max_score) { 
                    max_score = score;
                    best_pos = cell;
                }
            }
            return std::make_pair(max_score, best_pos);
        } 
        else {
            int min_score = 9999999;
            int best_pos = -1;
            for (int cell : empty_cells()) {
                board[cell] = 'X';
                auto result = minimax(cell, true);
                board[cell] = ' ';
                int score = result.first;
                if(score < min_score) {
                    min_score = score;
                    best_pos = cell;
                }
            }
            return std::make_pair(min_score, best_pos);
        }
    }

    void computerNextMove() {
        int cell;
        int temp_cell;
        int temp_score;
        int best_cell = 0;
        int best_score = -9999999;
        
        if(empty_cells().size() == 9) {
            best_cell = corner_cells[randomInt(0, corner_cells.size() - 1)];
        } 
        else {
            for (int cell : empty_cells()) {
                auto result    = minimax(cell, true);
                int temp_score = result.first;
                int temp_cell  = result.second;
                if (temp_score > best_score) {
                    best_score = temp_score;
                    best_cell  = temp_cell;
                }
            }

            played.insert(best_cell);
            board[best_cell] = 'O';
            gui.displayBoard(board);
        }
    }

    bool terminate(char who) {
        if (hasWon(who) && who == 'X') {
            gui.displayMessage("You won! Thanks for playing.");
            return true;
        } else if (hasWon(who) && who == 'O') {
            gui.displayMessage("You lost! Thanks for playing.");
            return true;
        } else if (played.size() == 9) {
            gui.displayMessage("A draw! Thanks for playing.");
            return true;
        } else {
            return false;
        }
    }

    void restart() {
        for (int i = 0; i < 9; i++) {
            board[i] = ' ';
        }
        played.clear();

        bool computer_goes_first = randomBool();

        if (computer_goes_first) {
            computer_first_turn = true;
            gui.displayMessage("Heads. I go first!");
            computerNextMove();
        } else {
            computer_first_turn = false;
            gui.displayMessage("Tails. you go first!");
        }

        gui.displayBoard(board);
    }

private:
    std::vector<char> board;
    std::set<int> played;
    std::vector<std::array<int,3>> win_sets;
    std::vector<int> corner_cells;
    bool computer_first_turn = false;
    int num_win_sets;

    Screen gui;
    
    // random gen
    bool randomBool() {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        std::uniform_int_distribution<int> dist(0,1);
        return dist(gen) == 1;
    }

    int randomInt(int min_val, int max_val) {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        std::uniform_int_distribution<int> dist(min_val, max_val);
        return dist(gen);
    }
};



int main() {
    TicTactoe game;
    Screen gui;
    while (1) {
        while (1) {
            game.playerNextMove();
            if(game.terminate('X')) break;
            game.computerNextMove();
            if(game.terminate('O')) break;
        }
        gui.displayGameStatus("Press [1] to continue game. Press anything else to end.");
        std::cout << "> ";
        std::string option;
        std::getline(std::cin, option);

        if (option == "1") {
            // Reinitialize the game
            // TicTactoe game;
            game.restart();
            continue;
        } else {
            break;
        }
    }
    return 0;
}